subsys/bluetooth/mesh/access.c

```diff
@@ -1002,6 +1006,69 @@ uint16_t *bt_mesh_model_find_group(struct bt_mesh_model **mod, uint16_t addr)
 	return ctx.entry;
 }
 
+#if CONFIG_BT_MESH_LABEL_COUNT > 0
+static const uint8_t **model_uuid_get(struct bt_mesh_model *mod, const uint8_t *uuid)
+{
+	int i;
+
+	for (i = 0; i < CONFIG_BT_MESH_LABEL_COUNT; i++) {
+		if (mod->uuids[i] == uuid) {
```

Comment:
`bt_mesh_model_find_uuid` is internal and is not intended to be used by customers. The idea is exactly to avoid `memcmp` here and in all other code as it is much more efficient to do this way. 

Allowing customers to use their uuid creates 2 issues. First, it would require them to keep the memory with uuid until the message is sent out to the air. This is because of the case where Friend node re-encrypts own message in Friend Queue targeted to LPN (see `unseg_app_sdu_prepare` in `friend.c`). Second, even if they are able to send the message using their own uuid, for receiving the message the transport layer still needs to know the uuid in advance. And in this case uuid should be added to the stack anyway. Which raises a question: why not add virtual address to the stack in advance and then use the pointer everywhere and avoid `memcmp`?

One more point to the design I think I should emphasize.

For customers now there are 2 places where they have ability to set virtual address as a destination: through `bt_mesh_msg` and `bt_mesh_model_pub` structures. The second one should not really be altered by application and should only be controlled by Configuration Client.

From the spec point of view there is no way to simply add or remove virtual address. Virtual address is only added through subscription or publication states. And when added, it also affects LPN as the node needs send a subscription message to its friend. When the address is removed, all models should be unsubscribed from it, as well as LPN.

I intentionally didn't leave a public API to add/delete/get virtual addresses for app, because the cases above need to be considered in such API, but the benefit is not clear for me. In addition to this, when the access layer receives a virtual address message, it should find a model subscribed to that address. So without the subscribing a model to the address through Configuration Client, usage of virtual address on its own makes less sense. We have a callback that can be used to process any received message (`CONFIG_BT_MESH_ACCESS_LAYER_MSG`). This is I guess they only case where it could potentially be used. I may add getter for the address, if you think it is necessary. But adding and removing should IMO be done through Configuration Client.

Reply:
It is difficult to say what is better for virtual address usage. I didn't see real use cases. Let's leave it `as is` if anybody claims any problem with it then it will be, at least, a direction for improvement. right now there is no reason to reach `perfect` solution.